### PR TITLE
Keystore Specification

### PR DESCRIPTION
**Pull request by CarlBeek**
This is a proposal for the keystore specification within Eth2.0. It is the same as Eth1 keystores, but with a narrowing of supported functions leading to a simpler implementation for clients.

### CODE REVIEW

File: keystore-specs/keystore.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+**Why not support PBKPRF2?**
+
+It is not that PBKPRF2 is bad (in fact it is a part of scrypt under the hood), it is just nice to present a simplified keystore with fewer versions so that client implementors need only support a single function.
+
+**Why are private keys encoded with Big Endian?**
+
+This is done because it is how keys are stored in Eth1 and because it is is the standard of most of the crypto libraries.
+
+**Why is Keccak used, doesn't it introduce another security assumption and dependency?**
```
Comment (ralexstokes) [inline keystore-specs/keystore.md]: i'd suggest using `sha-256` just for consistency, we may as well minimize a prior dependency where we can

File: keystore-specs/keystore.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+Eth2.0. will exclusively use scrypt as the Key Derivation Function for the private keys. This is a slight variation of the Eth1.0 keystores as scrypt MUST be supported and PBKDF2 MUST NOT be used as the KDF for a validator's signing or withdrawal keys.
+
+## A high-level overview of how Keystores work
+
+![Keystore](./keystore.png)
+
+## Definition
+
+The process of obtaining a keystore can be broken into 3 sub-sections: obtaining the *derived key*, checking the derived key's validity, and obtaining the public key.
```
Comment (ralexstokes) [inline keystore-specs/keystore.md]:
```suggestion
The process of obtaining a keystore can be broken into 3 sub-sections: obtaining the *derived key*, checking the derived key's validity, and obtaining the private key.
```

File: keystore-specs/keystore.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+```python
+derived_key = scrypt(password, dklen, n, p, r, salt)
+```
+
+### Checking the derived key's validity
+
+The derived key's validity is checked by determining whether the Keccak256-hash of the second 16 bytes (from the left) of the `derived_key` and the `ciphertext`, yields the `mac`.
+
+```python
+assert keccak256(derived_key[16:32] + ciphertext) == mac
```
Comment (ralexstokes) [inline keystore-specs/keystore.md]: see comment below about `sha-256`.

**Comment by ralexstokes** [comment]
some comments on the `keystore.png`:

1) suggestion: make it clearer that you get one output from scrypt which you then split into a "left" and "right" half.
2) you use the symbol commonly used for `xor` when you mean `concatenation` -- change it to make it clearer? one option is to use the concatentation symbol `||`, possibly in a circle (like how xor is a "plus in a circle")

otherwise it is a nice diagram :)

**Comment by CarlBeek** [comment]
@ralexstokes Thanks so much for your feedback on this!

I too would like to stick with SHA256 everywhere, but unfortunately in order to remain compatible with Eth1, this is simply not possible. (Hence the bad joke about "Keccak haunting Ethereum forever.")

I have subsequently decided that it is much cleaner just to come up with a new specification that addresses this and a few other concerns of mine. As a result, I am closing this in favour of https://github.com/ethereum/eth2.0-specs/pull/1361

**Comment by djrtwo** [comment]
Closed in favor of https://github.com/ethereum/eth2.0-specs/pull/1361

**Pull request closed without merging**